lidia/lidia.go

```diff
@@ -0,0 +1,261 @@
+	io.ReadCloser
+	io.ReaderAt
+}
+
+// Table represents a lidia symbol table that can be queried for lookups.
+type Table struct {
+	file ReaderAtCloser
+	hdr  header
+	opt  options
+
+	vaTable []byte
+
+	fieldsBuffer []byte
+}
+
+// SourceInfoFrame represents a single frame of symbolized profiling information.
+// It contains the name of the function, the source file path, and the line number
+// at which the profiling sample was taken.
+type SourceInfoFrame struct {
+	LineNumber   uint64
+	FunctionName string
+	FilePath     string
+}
+
+// GoRange represents a function range to be added to a lidia file.
+type GoRange struct {
+	VA        uint64
+	Length    uint32
+	Function  string
+	File      string
+	CallFile  string
+	CallLine  uint32
+	Depth     uint32
+	LineTable LineTable
+}
+
+// LineTable represents source line number information.
+type LineTable []LineTableEntry
+
+// LineTableEntry maps an offset to a line number.
+type LineTableEntry struct {
+	Offset     uint32
+	LineNumber uint32
+}
+
+// OpenReader creates a new Table from the provided ReaderAtCloser.
+// It reads the header, validates the format, and prepares the table for lookups.
+// The caller is responsible for closing the returned Table.
+func OpenReader(f ReaderAtCloser, opt ...Option) (*Table, error) {
+	var err error
+	res := new(Table)
+
+	for _, o := range opt {
+		o(&res.opt)
+	}
+
+	res.file = f
+
+	hdr, err := readHeader(f)
+	if err != nil {
+		res.Close()
+		return nil, fmt.Errorf("failed to read header: %w", err)
+	}
+
+	if hdr.magic != magic {
+		res.Close()
+		return nil, fmt.Errorf("invalid magic number: expected 0x%x, got 0x%x", magic, hdr.magic)
+	}
+
+	if hdr.version != version {
+		res.Close()
+		return nil, fmt.Errorf("unsupported version: expected %d, got %d", version, hdr.version)
+	}
+
+	if hdr.vaTableHeader.entrySize != 4 && hdr.vaTableHeader.entrySize != 8 {
+		res.Close()
+		return nil, fmt.Errorf("invalid vaSize: %d, expected 4 or 8", hdr.vaTableHeader.entrySize)
+	}
+
+	if hdr.rangeTableHeader.fieldSize != 4 && hdr.rangeTableHeader.fieldSize != 8 {
+		res.Close()
+		return nil, fmt.Errorf("invalid fieldSize: %d, expected 4 or 8", hdr.rangeTableHeader.fieldSize)
+	}
+
+	if hdr.rangeTableHeader.count != hdr.vaTableHeader.count {
+		res.Close()
+		return nil, fmt.Errorf("count mismatch: range table count (%d) != VA table count (%d)",
+			hdr.rangeTableHeader.count, hdr.vaTableHeader.count)
+	}
+
+	res.hdr = hdr
+
+	res.fieldsBuffer = make([]byte, int(hdr.rangeTableHeader.fieldSize)*fieldsCount)
+	// all functions addresses sorted.
+	res.vaTable = make([]byte, int(hdr.vaTableHeader.entrySize)*int(hdr.vaTableHeader.count))
+
+	if _, err = f.ReadAt(res.vaTable, int64(hdr.vaTableHeader.offset)); err != nil {
+		res.Close()
+		return nil, fmt.Errorf("failed to read VA table: %w", err)
+	}
+
+	if res.opt.crc {
+		if err = res.CheckCRC(); err != nil {
+			res.Close()
+			return nil, fmt.Errorf("CRC check failed: %w", err)
+		}
+	}
```

Comment:
Keep in mind that `hdr` might be completely invalid – we never validate its checksum (we don't even have it). It means that things like `make([]byte, int(hdr.rangeTableHeader.fieldSize)*fieldsCount)` may cause OOM, overflows, and all sorts of issues caused by bad data.

I'd like to see a fuzz test – let's not worry about this too much right now, but we should definitely add it.